feat(ootbc,hybrid): add guide how to run in hybrid mode

igpetrov opened the pull request:
## Description

Adding guide on how to run Connectors in hybrid mode.

## When should this change go live?

With `0.23.0` Connectors release - 12 Sept., 2023.

- [x] This change is not yet live and should not be merged until 12 Sept., 2023 (apply `hold` label or convert to draft PR)?
- [x] There is no urgency with this change.
- [ ] This change or page is part of a marketing blog, conference talk, or something else on a schedule.
- [ ] This functionality is already available but undocumented.
- [ ] This is a bug fix or security concern.

## PR Checklist

<!-- Keep in mind, Camunda maintains 18 months of versions. Backporting your change or including it in multiple versions is common. -->

- [x] I have added changes to the relevant `/versioned_docs` directory, or they are not for an **already released version**.
- [x] I have added changes to the main `/docs` directory (aka `/next/`), or they are not for **future versions**.
- [ ] My changes require an [Engineering review](https://github.com/camunda/camunda-platform-docs/blob/main/howtos/documentation-guidelines.md#review-process), and I've assigned an engineering manager or tech lead as a reviewer, or my changes do not require an Engineering review.
- [x] My changes require a [technical writer review](https://github.com/camunda/camunda-platform-docs/blob/main/howtos/documentation-guidelines.md#review-process), and I've assigned @christinaausley as a reviewer, or my changes do not require a technical writer review.

igpetrov commented:
Hi @bastiankoerber 

Please take a look at early version of the doc.

akeller commented:
@igpetrov does this close #2351?

akeller reviewed (dismissed):
Please move this into **Next steps** in the `sidebars.js` since the Guides section is currently split between the **Getting started guides** and **Next steps**.

sbuettner commented on docs/guides/use-connectors-in-hybrid-mode.md:
````diff
@@ -0,0 +1,116 @@
+Run the following script:
+
+```shell
+docker run --rm --name=HybridConnector \
````
```suggestion
docker run --rm --name=HybridConnectorRuntime \
```

sbuettner commented on docs/guides/use-connectors-in-hybrid-mode.md:
````diff
@@ -0,0 +1,116 @@
+
+Run the following script:
+
+```shell
+docker run --rm --name=HybridConnector \
+    -e ZEEBE_CLIENT_SECURITY_PLAINTEXT=false \
+    -e ZEEBE_CLIENT_CLOUD_CLUSTER-ID='<YOUR_CLUSTER_ID>' \
+    -e ZEEBE_CLIENT_CLOUD_CLIENT-ID='<YOUR_CLIENT_ID>' \
+    -e ZEEBE_CLIENT_CLOUD_CLIENT-SECRET='<YOUR_CLIENT_SECRET>' \
+    -e ZEEBE_CLIENT_CLOUD_REGION='<YOUR_CLUSTER_REGION>' \
+    -e CAMUNDA_OPERATE_CLIENT_URL='https://<region>.operate.camunda.io/<cluster-id>' \
+    -e CONNECTOR_HTTP_REST_TYPE='io.camunda:http-json:2' \
+        camunda/connectors-bundle:<desired-version>
+```
````
We dont need to repear this as the script is the same as above.

igpetrov replied:
Done. Reworded with reference.

sbuettner commented on docs/guides/use-connectors-in-hybrid-mode.md:
```diff
@@ -0,0 +1,116 @@
+
+### Explanation
+
+Pay attention at the line `-e CONNECTOR_HTTP_REST_TYPE='io.camunda:http-json:2'`. This line means that we have to override
+`CONNECTOR_X_TYPE` with given type. In this case, we want to register local self-managed HTTP REST connector as `io.camunda:http-json:2`.
+The `X` is normalized to environment variable Connector name. For example, the [HTTP REST Connector](https://github.com/camunda/connectors/blob/main/connectors/http/rest/src/main/java/io/camunda/connector/http/rest/HttpJsonFunction.java#L33)
+`HTTP REST` name becomes `HTTP_REST`, or the [Kafka Consumer Connector](https://github.com/camunda/connectors/blob/main/connectors/kafka/src/main/java/io/camunda/connector/kafka/inbound/KafkaExecutable.java#L20) name
+transformed to `KAFKA_CONSUMER`, therefore to override it one would need to pass `CONNECTOR_KAFKA_CONSUMER_TYPE=xxx` environment variable.
```
```suggestion
Pay attention at the line `-e CONNECTOR_HTTP_REST_TYPE='io.camunda:http-json:local'`. This line means that we have to override
`CONNECTOR_X_TYPE` with given type. In this case, we want to register local self-managed HTTP REST connector as `io.camunda:http-json:local`.
The `X` is normalized to environment variable Connector name. For example, the [HTTP REST Connector](https://github.com/camunda/connectors/blob/main/connectors/http/rest/src/main/java/io/camunda/connector/http/rest/HttpJsonFunction.java#L33)
`HTTP REST` name becomes `HTTP_REST`, or the [Kafka Consumer Connector](https://github.com/camunda/connectors/blob/main/connectors/kafka/src/main/java/io/camunda/connector/kafka/inbound/KafkaExecutable.java#L20) name
transformed to `KAFKA_CONSUMER`, therefore to override it one would need to pass `CONNECTOR_KAFKA_CONSUMER_TYPE=xxx` environment variable.
```

igpetrov commented:
Hi @akeller, this PR is currently pre-review, because the topic is complex and I wanted my team have a look first. All version docs, sidebars, etc will be added but later.

I didn't know that the ticket https://github.com/camunda/camunda-platform-docs/issues/2351 exist, but it addresses exactly it.

igpetrov commented on docs/guides/use-connectors-in-hybrid-mode.md:
```diff
@@ -0,0 +1,116 @@
+    -e ZEEBE_CLIENT_CLOUD_CLIENT-SECRET='<YOUR_CLIENT_SECRET>' \
+    -e ZEEBE_CLIENT_CLOUD_REGION='<YOUR_CLUSTER_REGION>' \
+    -e CAMUNDA_OPERATE_CLIENT_URL='https://<region>.operate.camunda.io/<cluster-id>' \
+    -e CONNECTOR_HTTP_REST_TYPE='io.camunda:http-json-local' \
```
```suggestion
    -e CONNECTOR_HTTP_REST_TYPE='io.camunda:http-json:local' \
```

igpetrov commented on docs/guides/use-connectors-in-hybrid-mode.md:
```diff
@@ -0,0 +1,116 @@
+    -e ZEEBE_CLIENT_CLOUD_CLIENT-SECRET='<YOUR_CLIENT_SECRET>' \
+    -e ZEEBE_CLIENT_CLOUD_REGION='<YOUR_CLUSTER_REGION>' \
+    -e CAMUNDA_OPERATE_CLIENT_URL='https://<region>.operate.camunda.io/<cluster-id>' \
+    -e CONNECTOR_HTTP_REST_TYPE='io.camunda:http-json:2' \
```
```suggestion
    -e CONNECTOR_HTTP_REST_TYPE='io.camunda:http-json:local' \
```

bastiankoerber commented on docs/guides/use-connectors-in-hybrid-mode.md:
```diff
@@ -0,0 +1,103 @@
+
+To name few use-cases where this approach might be useful:
+
+- When you deal with business critical secrets, such as financial information, personal identifiable information, other information that is legally not allowed to share with Camunda SaaS clusters.
```
i would remove this one, this may confuses the reader or will scare him :-)

igpetrov replied:
Okay, I'll rephrase this one. What I mean here is a typical Kafka/RabbitMQ experience, where broker (Kafka, RabbitMQ) is a backbone for dealing with mission-critical data and typically not exposed to the public internet.

igpetrov replied:
Done

bastiankoerber commented on docs/guides/use-connectors-in-hybrid-mode.md:
```diff
@@ -0,0 +1,103 @@
+- Publish new element template.
+- Use new published template in the BPMN diagram as usual.
+
+For example, if you defined `CONNECTOR_HTTP_REST_TYPE='io.camunda:http-json:local'` in runtime, you have to do the following in the
```
should we describe here the options the have:
making the id visible
Adding a dropdown for the type (run in SM and run in SAAS)

At the end it is totally up to the user, but i think it would be valuable to show here the options.
- duplicating template and hiding the id
- Adding dropdown to a REST
- make it visible and collapse the group

igpetrov replied:
If you think, it's valuable, we can add it. For now, we pre-created hybrid examples (see appendix section), but can also describe those options too.

igpetrov replied:
Done

bastiankoerber reviewed (commented):
i really like the guides. 
just have to minor suggestions

igpetrov commented:
@akeller this one is ready for review

akeller commented:
@christinaausley can you review this? I'm hoping to get around to reviewing it later today or tomorrow.

For complete transparency, we won't be able to walk through this and validate the steps are correct.

christinaausley commented on docs/guides/use-connectors-in-hybrid-mode.md:
```diff
@@ -0,0 +1,163 @@
+For example, see a relation between [Kafka element template](https://github.com/camunda/connectors/blob/main/connectors/kafka/element-templates/kafka-inbound-connector.json#L39)
+and [Kafka Connector](https://github.com/camunda/connectors/blob/main/connectors/kafka/src/main/java/io/camunda/connector/kafka/inbound/KafkaExecutable.java#L20).
+
+To make hybrid Connector runtime work correctly, it is required to override Connector type.
```
Do you mean "For the hybrid Connectors runtime to work properly, you must override the Connector type."?

Also note, I have an open issue regarding "Connector runtime" or "Connectors" runtime because we use this inconsistently. https://github.com/camunda/developer-experience/issues/113

igpetrov replied:
That's correct. Fixed it.

christinaausley commented on docs/guides/use-connectors-in-hybrid-mode.md:
```diff
@@ -0,0 +1,163 @@
+Use this option when you plan to clearly indicate that a specific Connector will only be used in a specific use-case.
+Otherwise, users might be confused between two of the same Connector types.
+
+For example, if you defined `CONNECTOR_HTTP_REST_TYPE='io.camunda:http-json:local'` in runtime, you must implement the following in the
```
In Connectors runtime, specifically here? May need to add this to my Connector(s) runtime issue.

igpetrov replied:
reworded a bit

christinaausley reviewed (dismissed):
Technical review in latest commit and a few comments for clarity 👍

akeller commented:
@igpetrov can you have a look at the changes requested so we can get this merged for tomorrow's release?

igpetrov commented:
Hi @akeller @christinaausley I completely missed that there were comments, sorry about it!

I reworded/addressed issues, and removed `hold` label. This one is ready to be merged.

Thank you!
